Variable buffer size

AndrzejKurek opened the pull request:
This PR adds a buffer resizing feature before and after the handshake if `MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH` is defined. The work is based on https://github.com/ARMmbed/mbedtls/pull/3074.
All of the handshake tests now have buffer resizing on by default, so thanks to `component_test_variable_ssl_in_out_buffer_len` in `all.sh` are performed with this option turned on.
Darryl's changes to the I/O buffer lengths had to be split between ssl files, because of the changes on `development` .

Patater commented on library/ssl_tls.c:
```diff
@@ -5818,6 +5863,56 @@ void mbedtls_ssl_handshake_free( mbedtls_ssl_context *ssl )
+    }
+    if( modified )
+    {
+        mbedtls_ssl_reset_in_out_pointers( ssl );
```
Why is it important to wait until here to call mbedtls_ssl_reset_in_out_pointers()? Add a C comment to explain the efficiency gained by waiting until here.

AndrzejKurek replied:
It's just a small improvement to not call `mbedtls_ssl_reset_in_out_pointers` twice, since the changed values won't be used between the first and second buffer resizes. It seems pretty simple, how would you like me to explain it?
`Update pointers here to avoid doing it twice.`?

Patater commented on library/ssl_tls.c:
```diff
@@ -5818,6 +5863,56 @@ void mbedtls_ssl_handshake_free( mbedtls_ssl_context *ssl )
+        {
+            memcpy( resized_buffer, ssl->out_buf, buf_len );
+            mbedtls_platform_zeroize( ssl->out_buf, ssl->out_buf_len );
+            mbedtls_free( ssl->out_buf );
```
Do we really need to preserve the contents of the buffer here? What is present? Please add a C comment justifying the preservation of the data.

AndrzejKurek replied:
Quoting Janos: 

> Mbed TLS processes TLS records one by one and does not have more than a single
> record in the I/O buffers at a given time. When the handshake completes and the
> resizing happens it is safe to assume that both buffers are empty.
> 
> On the other hand, it processes incoming DTLS data datagram by datagram. This
> means that the input buffer needs to be able to hold a whole datagram. A
> datagram is allowed to have several DTLS records in it.
> 
> This has two consequences for DTLS:
> 
> 1. after the handshake it is not safe to assume anymore that the I/O buffers are
>    empty
> [...]
Should I add a comment: `Because of the way mbedTLS processes datagrams and the fact that a datagram is allowed to have several records in it, it is possible that the I/O buffers are not empty at this stage`?

Patater commented on library/ssl_tls.c:
```diff
@@ -3739,18 +3741,24 @@ int mbedtls_ssl_setup( mbedtls_ssl_context *ssl,
     if( ssl->out_buf == NULL )
     {
-        MBEDTLS_SSL_DEBUG_MSG( 1, ( "alloc(%d bytes) failed", MBEDTLS_SSL_OUT_BUFFER_LEN) );
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "alloc(%d bytes) failed", out_buf_len) );
```
Pre-existing style issue, but we may as well fix it:
```diff
- out_buf_len) );
+ out_buf_len ) );
```

Patater commented on include/mbedtls/config.h:
```diff
@@ -1779,6 +1779,13 @@
  */
```
Commit "Add I/O buffer length fields to mbedtls_ssl_context" needs Darryl's signed-off-by as he was the primary author. You should have your signed-off-by as well if you changes the code, for a total of two signed-off-by lines. Same goes for the other commit from Darryl.

AndrzejKurek replied:
Should I sign it off for @dgreen-arm? Or should he prepare such commit?

Patater commented on tests/scripts/all.sh:
```diff
@@ -1142,6 +1142,19 @@ component_test_no_max_fragment_length_small_ssl_out_content_len () {
+
+    msg "test: compat.sh, MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH enabled"
+    if_build_succeeded tests/compat.sh
+}
```
In commit "Add variable buffer length test to all.sh", we should probably do variable buffer length both with and without CID enabled, not just without CID. Maybe the config full test does this already: please check.

AndrzejKurek replied:
Yes it does, that was the test that failed previously, but I would like to add a unit test to the ssl test suite doing such test.

Patater commented on include/mbedtls/ssl_internal.h:
```diff
@@ -256,6 +256,22 @@
       + ( MBEDTLS_SSL_CID_OUT_LEN_MAX ) )
 #endif
 
+#if defined(MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH)
```
In commit "Add static functions to calculate ssl buffer sizes ", we say "Add two, so far unused functions to calculate the input and output ssl buffer sizes." Instead, explain what they'll be used for if they deserve to be added in their own commit, or move their addition to the commit that uses them (that commit would also contain their justification already).

Patater commented on include/mbedtls/ssl_internal.h:
```diff
@@ -256,6 +256,22 @@
+{
+    return (uint32_t) mbedtls_ssl_get_max_frag_len( ctx )
+               + MBEDTLS_SSL_HEADER_LEN + MBEDTLS_SSL_PAYLOAD_OVERHEAD
+               + MBEDTLS_SSL_CID_OUT_LEN_MAX;
```
In commit "Add static functions to calculate ssl buffer sizes", we probably don't want to add MBEDTLS_SSL_CID_OUT_LEN_MAX unless CID is configured. Or, we should define MBEDTLS_SSL_CID_OUT_LEN_MAX to 0 when CID is not configured and avoid adding any ifdefs here or later.

AndrzejKurek replied:
I'll move my fix to this commit.

Patater commented on library/ssl_tls.c:
```diff
@@ -3643,6 +3643,55 @@ static int ssl_handshake_init( mbedtls_ssl_context *ssl )
     }
+#if defined(MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH)
+    /* If the buffers are too small - reallocate */
+    {
```
Consider making a new static function rather than just a new block. It should help readability if we know exactly which items from the parent block scope are used, so we know what to consider and think about when reading the code.

Maybe the function could be named something like `reallocate_if_the_buffers_are_too_small()` :p

AndrzejKurek replied:
The problem is that both chunks use different conditions. If this was a different language, I'd use a lambda, but here - do you have an idea on how to do it in a clean, and not overcomplicated (like with pointers to functions passed) way?

gilles-peskine-arm replied:
The way I'd do it is, start by passing a boolean indicating whether to use the “too small” or “too large” check, and once that works see if a simplification becomes apparent.

Patater commented on library/ssl_tls.c:
```diff
@@ -5818,6 +5867,71 @@ void mbedtls_ssl_handshake_free( mbedtls_ssl_context *ssl )
+
+#if defined(MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH)
+    /* If the buffers are too big - reallocate */
+    {
```
Consider making a new static function rather than a block.

Patater commented on include/mbedtls/ssl_internal.h:
```diff
@@ -238,7 +238,7 @@
 #define MBEDTLS_SSL_HEADER_LEN 13
 
-#if defined(MBEDTLS_SSL_DTLS_CONNECTION_ID)
+#if !defined(MBEDTLS_SSL_DTLS_CONNECTION_ID)
```
In commit "Fix I/O buffer sizes when MBEDTLS_SSL_DTLS_CONNECTION_ID is enabled", describe in the git commit message, on its own line, that the commit "Fixes #3077"

Patater replied:
In commit "Fix I/O buffer sizes when MBEDTLS_SSL_DTLS_CONNECTION_ID is enabled", we say "Update related buffer size calculations.", but there is no such code change in the patch. Also, consider making the commit early in your patch set (could be the first commit in this set, as it just fixes a pre-existing bug).

Patater commented on library/ssl_tls.c:
```diff
@@ -3643,6 +3643,55 @@ static int ssl_handshake_init( mbedtls_ssl_context *ssl )
+                mbedtls_platform_zeroize( ssl->in_buf, ssl->in_buf_len );
+                mbedtls_free( ssl->in_buf );
+
+                ssl->in_buf = resized_buffer;
```
Does LCOV show we reached here in our unit testing?

AndrzejKurek replied:
Yes.

Patater commented on library/ssl_tls.c:
```diff
@@ -5818,6 +5867,71 @@ void mbedtls_ssl_handshake_free( mbedtls_ssl_context *ssl )
+                mbedtls_platform_zeroize( ssl->in_buf, ssl->in_buf_len );
+                mbedtls_free( ssl->in_buf );
+
+                ssl->in_buf = resized_buffer;
```
Does LCOV show we reached here in our unit testing?

AndrzejKurek replied:
Good suggestion - while checking that I've found that i have decreased test coverage by doing the following:
- First, I had separate 'integration' tests doing every single feature + variable buffers and serialization + MFL + variable buffers, renegotiation + MFL + variable buffers.
- Then I figured that if we don't want to duplicate so many tests just to add the variable buffers, we could run the whole test suite with variable buffers turned on by default.
- I removed the additional tests doing every single feature + variable buffers, but by doing so, I also removed the `integration` tests - MFL + serialization + VB, MFL + renegotiation + VB.

I'll add these tests as a part of the last commit.

AndrzejKurek commented:
I would also like to add two tests to this PR - CID + MFL + VB with max data lengths, for various MFL's, and a record splitting test which will fail without the two additional lines updating in and out msg pointers.

AndrzejKurek commented:
I have added a bunch of updates:
- Added CID and CBC record splitting tests to all.sh;
- Forced a record splitting ciphersuite with MFL handshake fragmentation tests (these were failing without the in and out msg pointer updates, so now we have coverage of it in unit tests);
- Added comments;
- Added sign-offs;
- Added Piotr's memory acceptance test (passing now), and tweaked it to use VARIABLE_BUFFER_LENGTHS define;
- Added tests in test suites integrating MFL + serialization + VB, MFL + renegotiation + VB (to cover the buffer expansion code too) ;
- Merged the static function commit with the commit that actually used it.

Patater commented on library/ssl_tls.c:
```diff
@@ -5818,6 +5868,75 @@ void mbedtls_ssl_handshake_free( mbedtls_ssl_context *ssl )
                               sizeof( mbedtls_ssl_handshake_params ) );
+
+#if defined(MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH)
+    /* If the buffers are too big - reallocate. Because of the way mbedTLS
```
```diff
- mbedTLS
+ Mbed TLS
```

AndrzejKurek commented:
Introduced requested changes and fixed one missing dependency.

Patater commented:
CI failure is Mbed OS import of Mbed TLS, which looks for the recently removed `apache-2.0.txt` file. The `pr-head` job passes, so we know this PR works fine with Mbed OS. This means CI has passed.

Patater commented on programs/ssl/ssl_server2.c:
```diff
@@ -3742,6 +3745,13 @@ int main( int argc, char *argv[] )
 #endif /* MBEDTLS_SSL_DTLS_CONNECTION_ID */
 
+#if defined(MBEDTLS_MEMORY_DEBUG)
+    mbedtls_memory_buffer_alloc_cur_get(&current_heap_memory, &heap_blocks);
```
Style: Need spaces after `(` and before `)`, here and other places in commit "Add an acceptance test for memory usage after handshake"

AndrzejKurek commented:
Added the suggested static function and fixed spaces in Piotr's commit.

AndrzejKurek commented:
Also, I think it would be good to request a review from the OSS team.

AndrzejKurek commented:
One line fix (and a comment) pushed - changed the size of buffer that is being copied inside the new static function.

AndrzejKurek commented:
The only failures seem to be `cp: cannot stat 'TARGET_IGNORE/mbedtls/apache-2.0.txt': No such file or directory`.

Patater commented on library/ssl_tls.c:
```diff
@@ -245,6 +245,25 @@ int mbedtls_ssl_session_copy( mbedtls_ssl_session *dst,
+        return -1;
+
+    /* Copy the smaller one out of old and new buffer lengths. */
+    memcpy( resized_buffer, *buffer, ( len < *len_info ) ? len : *len_info );
```
I'm not sure when we would use `*len_info` as the length of the copy, or if the previous `mbedtls_calloc()` might be too large if it is always `len`. Could you expand a bit more in the C comment, please?

AndrzejKurek replied:
When downsizing the buffer, len is smaller, and when upsizing - len_info. I'll expand the comment.

AndrzejKurek replied:
Done.

Patater replied:
Yes, I should consider this function used in both ways. OK. Much clearer now in the code. Thanks.

AndrzejKurek commented:
The following tests fail:

AndrzejKurek commented:
The following tests fail:
- DTLS client reconnect from same port: reconnect;
- K64F due to the apache file fail.

AndrzejKurek commented:
@Patater - these are pre-existing and I would like to fix them in a separate PR, since they're from lower level, message queue tests.

AndrzejKurek commented:
Fixed here: https://github.com/ARMmbed/mbedtls/pull/3092

Patater commented:
> @Patater - these are pre-existing and I would like to fix them in a separate PR, since they're from lower level, message queue tests.

OK, we've merged #3092 now. Next time we run the merge tests with gcc 7, we should be good.

Patater commented:
I've kicked off the merge CI job.

Patater commented:
CI passed on merge job. Good for merging.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -1142,6 +1142,75 @@ component_test_no_max_fragment_length_small_ssl_out_content_len () {
+    if_build_succeeded tests/compat.sh
+}
+
+component_test_variable_ssl_in_out_buffer_len_CID () {
```
Thanks for offering to make a follow-up. You don't have to, this is pretty minor. [I've recorded as something to review one day](https://github.com/ARMmbed/mbedtls/issues/3417). I was asking here in case there was something I'd missed.
